### PR TITLE
Assessment form

### PR DESCRIPTION
**Pull request by jaleszek**
https://podio.com/bitzesty/app-qae/apps/stories/items/93
https://podio.com/bitzesty/app-qae/apps/stories/items/91 - appraisal form part
https://podio.com/bitzesty/app-qae/apps/stories/items/90
https://podio.com/bitzesty/app-qae/apps/stories/items/89

AssessorAssignment stores:
1. information about assessor assignment - assessor_id
2. position (secondary,primary) - position
3. document with assessment - document

Document is stored as hstore and has set up the accessors in accordance with the type of the form answer.

### CODE REVIEW

File: app/assets/javascripts/admin/form_answers.js.coffee [hunk 1]
```diff
@@ -29,6 +33,12 @@ ready = ->
   $(document).on "click", ".form-edit-link", (e) ->
     e.preventDefault()
     $(this).closest(".form-group").addClass("form-edit")
+  $(".submit-assessment").on "ajax:error", (e, data, status, xhr)->
```
Comment (matthewford) [inline app/assets/javascripts/admin/form_answers.js.coffee]: Function arrow (->) must be spaced properly

File: app/assets/javascripts/admin/form_answers.js.coffee [hunk 2]
```diff
@@ -29,6 +33,12 @@ ready = ->
...
+    errors = data.responseJSON
+    $(this).find(".feedbackHolder").html(errors.error)
+  $(".submit-assessment").on "ajax:success", (e, data, status, xhr)->
```
Comment (matthewford) [inline app/assets/javascripts/admin/form_answers.js.coffee]: Function arrow (->) must be spaced properly

File: app/assets/javascripts/admin/form_answers.js.coffee [hunk 3]
```diff
@@ -62,5 +73,24 @@ editFormAnswerAutoUpdate = ->
...
+  $(document).on "click", "#{klass} .btn-rag .dropdown-menu a", (e) ->
+    e.preventDefault()
+    ragClicked = $(this).closest("li").attr("class")
+    ragClicked = ragClicked.replace("rag-", "")
+    ragSection = $(this).parents(".form-group")
+    ragSection.find("option").each ->
+      if $(this).val() == ragClicked
+        $(this).parents("select").val(ragClicked)
+    $(klass).submit()
+bindAppraisalFields=(klass)->
```
Comment (matthewford) [inline app/assets/javascripts/admin/form_answers.js.coffee]: Function arrow (->) must be spaced properly

File: app/controllers/assessor/assessment_submissions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+class Assessor::AssessmentSubmissionsController < Assessor::BaseController
+  def create
```
Comment (matthewford) [inline app/controllers/assessor/assessment_submissions_controller.rb]: Assignment Branch Condition size for create is too high. [16.06/15]

File: app/models/assessor_assignment.rb [hunk 5]
```diff
@@ -2,24 +2,132 @@ class AssessorAssignment < ActiveRecord::Base
...
+
+  def primary?
+    position == PRIMARY_POSITION
+  end
+
+  def secondary?
+    position == SECONDARY_POSITION
+  end
+
+  def visible_for?(subject)
```
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Assignment Branch Condition size for visible_for? is too high. [17.49/15]

File: app/models/assessor_assignment.rb [hunk 6]
```diff
@@ -2,24 +2,132 @@ class AssessorAssignment < ActiveRecord::Base
   PRIMARY_POSITION = 0
   SECONDARY_POSITION = 1
 
-  validates :form_answer_id,
-            :assessor_id,
-            presence: true
+  begin :validations
+    validates :form_answer_id,
```
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb [hunk 7]
```diff
@@ -2,24 +2,132 @@ class AssessorAssignment < ActiveRecord::Base
...
+              presence: true
 
-  validates :position,
-            inclusion: {
-              in: [PRIMARY_POSITION, SECONDARY_POSITION]
-            },
-            presence: true
+    validates :position,
```
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb [hunk 8]
```diff
@@ -2,24 +2,132 @@ class AssessorAssignment < ActiveRecord::Base
...
+              inclusion: {
+                in: [PRIMARY_POSITION, SECONDARY_POSITION]
+              },
+              presence: true
 
-  belongs_to :assessor
-  belongs_to :form_answer
+    validate :award_specific_attributes
```
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb [hunk 9]
```diff
@@ -2,24 +2,132 @@ class AssessorAssignment < ActiveRecord::Base
...
+    validate :mandatory_fields_for_submitted
```
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb [hunk 10]
```diff
@@ -2,24 +2,132 @@ class AssessorAssignment < ActiveRecord::Base
...
+
+    validate do
```
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb [hunk 11]
```diff
@@ -2,24 +2,132 @@ class AssessorAssignment < ActiveRecord::Base
...
+      validate_rate :rag
+      validate_rate :strengths
+      validate_rate :verdict
+    end
+
+    validate :submitted_at_immutability
```
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb [hunk 12]
```diff
@@ -2,24 +2,132 @@ class AssessorAssignment < ActiveRecord::Base
...
+  end
+
+  begin :associations
+    belongs_to :assessor
```
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb [hunk 13]
```diff
@@ -2,24 +2,132 @@ class AssessorAssignment < ActiveRecord::Base
...
+    belongs_to :form_answer
```
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb [hunk 14]
```diff
@@ -2,24 +2,132 @@ class AssessorAssignment < ActiveRecord::Base
...
+    belongs_to :editable, polymorphic: true
```
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: spec/factories/assessors.rb [hunk 15]
```diff
@@ -6,8 +6,12 @@
...
 
-    trait :lead do
-      role "lead"
+    trait :lead_for_all do
+      trade_role "lead"
+      innovation_role "lead"
+      development_role "lead"
+      promotion_role "lead"
     end
+
```
Comment (matthewford) [inline spec/factories/assessors.rb]: Extra empty line detected at block body end.

File: app/controllers/assessor/assessment_submissions_controller.rb (hunk 4 above)
Comment (matthewford) [inline app/controllers/assessor/assessment_submissions_controller.rb]: Assignment Branch Condition size for create is too high. [16.06/15]

File: app/models/assessor_assignment.rb (hunk 5 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Assignment Branch Condition size for visible_for? is too high. [17.49/15]

File: app/models/assessor_assignment.rb (hunk 6 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 7 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 8 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 9 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 10 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 11 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 12 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 13 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 14 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/controllers/assessor/assessment_submissions_controller.rb (hunk 4 above)
Comment (matthewford) [inline app/controllers/assessor/assessment_submissions_controller.rb]: Assignment Branch Condition size for create is too high. [16.06/15]

File: app/models/assessor_assignment.rb (hunk 5 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Assignment Branch Condition size for visible_for? is too high. [17.49/15]

File: app/models/assessor_assignment.rb (hunk 6 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 7 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 8 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 9 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 10 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 11 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 12 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 13 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

File: app/models/assessor_assignment.rb (hunk 14 above)
Comment (matthewford) [inline app/models/assessor_assignment.rb]: Inconsistent indentation detected.

**Comment by RuslanHamidullin** [comment]
Good to merge :+1:

**Pull request merged**